File: platforms/android/stato/src/main/java/com/facebook/stato/nativeplugins/table/TableMetadata.java
package com.facebook.stato.nativeplugins.table;

import androidx.annotation.Nullable;

public class TableMetadata {

  final Column[] mColumns;
  final QueryableTableRowProvider mResponder;

  public QueryableTableRowProvider getResponder() {
    return mResponder;
  }

  private TableMetadata(
      @Nullable Column[] columns, @Nullable QueryableTableRowProvider queryResponder) {
    this.mColumns = columns == null ? new Column[] {} : columns;
    this.mResponder = queryResponder;
  }

  public static class Builder {
    private Column[] columns;
    private QueryableTableRowProvider queryResponder;

    public Builder columns(Column... columns) {
      this.columns = columns;
      return this;
    }

    public Builder queryResponder(QueryableTableRowProvider responder) {
      this.queryResponder = responder;
      return this;
    }

    public TableMetadata build() {
      return new TableMetadata(columns, queryResponder);
    }
  }
}

File: platforms/android/android-stato/src/main/kotlin/org/stato/nativeplugins/table/TableMetadata.kt
package org.stato.nativeplugins.table

class TableMetadata(
  columns: List<Column>,
  val responder: QueryableTableRowProvider?
) {

  val columns = columns.toList()

  class Builder {
    private var columns: List<Column> = emptyList()
    private var queryResponder: QueryableTableRowProvider? = null

    fun columns(vararg columns: Column): Builder {
      this.columns = this.columns + listOf(*columns)
      return this
    }

    fun queryResponder(responder: QueryableTableRowProvider): Builder {
      this.queryResponder = responder
      return this
    }

    fun build(): TableMetadata {
      return TableMetadata(columns, queryResponder)
    }
  }
}
